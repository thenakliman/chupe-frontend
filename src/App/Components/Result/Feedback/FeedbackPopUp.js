import React from 'react';
import propTypes from 'prop-types';

require('./FeedbackPopUp.css');

export class FeedbackPopUp extends React.Component {
  constructor(props) {
    super(props);
    this.state = {description: '', feedbackGivenToUser: ''};
    this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
    this.handleFeedbackTargetChange = this.handleFeedbackTargetChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  handleDescriptionChange(value) {
    this.setState({description: value});
  }

  handleFeedbackTargetChange(value) {
    this.setState({feedbackGivenToUser: value});
  }

  onSubmit() {
    this.props.onSubmit({
      description: this.state.description,
      givenTo: this.state.feedbackGivenToUser,
    });
  }

  render() {
    return (
        <div className={'feedback-pop-up-container'}>
          <div className={'feedback-pop-up-inner'}>
            <div>
              Feedback:
            </div>
            <textarea className={'feedback-description'}
                      value={this.state.feedback}
                      onChange={(event) => this.handleDescriptionChange(event.target.value)}
                      resize={'none'}/>

            Feedback To:
            <select value={this.state.feedbackGivenToUser}
                    className={'select-user-feedback'}
                    onChange={(event) => this.handleFeedbackTargetChange(event.target.value)}
            >
              <option key={''} value={''}> Select user</option>
              {
                this.props.users.map((user) =>
                    <option key={user.userName} value={user.userName}>
                      {user.userName}
                    </option>)
              }
            </select>
            <div className={'button-container'}>
              <button className={'feedback-buttons'}
                      id={'feedback-submit-button'}
                      onClick={this.onSubmit}>
                Submit
              </button>
              <button className={'feedback-buttons'}
                      id={'feedback-cancel-button'} onClick={this.props.onCancel}>
                Cancel
              </button>
            </div>
          </div>
        </div>
    );
  }
}

FeedbackPopUp.propTypes = {
  users: propTypes.array.isRequired,
  onSubmit: propTypes.func.isRequired,
  onCancel: propTypes.func.isRequired,
};
