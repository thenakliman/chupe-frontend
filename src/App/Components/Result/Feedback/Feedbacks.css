.feedback-container {
  margin: 1rem 18rem;
}

.give-feedback-button-container {
  display: flex;
  justify-content: flex-end;
  height: 3rem;
  margin-bottom: 1rem;
}

.give-feedback-button {
  cursor: pointer;
  padding: 0 1rem;
}