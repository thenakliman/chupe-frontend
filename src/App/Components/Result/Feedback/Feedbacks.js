import React from 'react';
import propTypes from 'prop-types';
import {FeedbackPopUp} from './FeedbackPopUp';

require('./Feedbacks.css');

export class Feedbacks extends React.Component {
    constructor(props) {
      super(props);
      this.state = {isGivingFeedback: false};
      this.toggleGiveFeedbackPopUp = this.toggleGiveFeedbackPopUp.bind(this);
      this.onSubmitOfFeedback = this.onSubmitOfFeedback.bind(this);
    }

    componentDidMount() {
      this.props.getAllFeedbacks(this.props.match.params.id);
      this.props.getAllUsers();
    }

    toggleGiveFeedbackPopUp() {
      this.setState({isGivingFeedback: !this.state.isGivingFeedback});
    }

    onSubmitOfFeedback(feedback) {
      this.props.giveFeedback({...feedback, sessionId: this.props.match.params.id});
      this.toggleGiveFeedbackPopUp();
    }

    render() {
      return (
         <div className={'feedback-container'}>
           {this.state.isGivingFeedback &&
              <FeedbackPopUp users={this.props.users}
                             onSubmit={this.onSubmitOfFeedback}
                             onCancel={this.toggleGiveFeedbackPopUp}
              />
           }
           <div className={'give-feedback-button-container'}>
             <button id='give-feedback-button'
                     className={'give-feedback-button'}
                     onClick={() => this.toggleGiveFeedbackPopUp()}>
                Give Feedback
             </button>
           </div>
           <table id='all-feedbacks-ordered-list'>
             <thead>
               <tr className='feedbacks-table-header'>
                 <th> Id </th>
                 <th> description </th>
                 <th> Given By </th>
               </tr>
             </thead>
             <tbody id='all-feedbacks-table-body-id'
                    className={'feedbacks-table-body'}>
             {
               this.props.feedbacks.map((feedback) => (
                 <tr key={`${feedback.id}`}>
                   <td> {feedback.id} </td>
                   <td className='feedbacks-row'>
                           {feedback.description}
                   </td>
                   <td> {feedback.givenBy}</td>
                 </tr>
               ))
             }
             </tbody>
           </table>
         </div>
      );
    }
}

Feedbacks.propTypes = {
  getAllFeedbacks: propTypes.func.isRequired,
  getAllUsers: propTypes.func.isRequired,
  giveFeedback: propTypes.func.isRequired,
  feedbacks: propTypes.array.isRequired,
  users: propTypes.array.isRequired,
};
