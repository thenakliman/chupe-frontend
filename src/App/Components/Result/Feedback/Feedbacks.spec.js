/* eslint-disable */
import React from 'react';
import {Feedbacks} from './Feedbacks';
import {FeedbackPopUp} from './FeedbackPopUp';
/* eslint-enable */
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';


const feedbacks = [{
  id: 10101,
  description: 'retro - 1',
}, {
  id: 10102,
  description: 'retro - 2',
}];

describe('Show feedback component', () => {
  describe('should match snapshot', () => {
    const wrapper = shallow(<Feedbacks
        match={{params: {id: 101}}}
        feedbacks={feedbacks}
        getAllFeedbacks={jest.fn()}
        getAllUsers={jest.fn()}
        giveFeedback={jest.fn()}
    />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call getAllFeedbacks on component did mount ', () => {
    const getAllFeedbacks = jest.fn();
    const feedbackSessionId = 101;
    shallow(
        <Feedbacks
            match={{params: {id: feedbackSessionId}}}
            feedbacks={feedbacks}
            getAllFeedbacks={getAllFeedbacks}
            getAllUsers={jest.fn()}
            giveFeedback={jest.fn()}
        />
    );

    expect(getAllFeedbacks).toHaveBeenCalledWith(feedbackSessionId);
  });

  it('should show row for each feedbacks', () => {
    const getAllFeedbacks = jest.fn();
    const feedbackSessionId = 101;
    const wrapper = shallow(
        <Feedbacks
            match={{params: {id: feedbackSessionId}}}
            feedbacks={feedbacks}
            getAllFeedbacks={getAllFeedbacks}
            getAllUsers={jest.fn()}
            giveFeedback={jest.fn()}
        />
    );

    expect(wrapper.find('tr').length).toBe(3);
  });

  it('should have initial state isGivingFeedback set to false', () => {
    const getAllFeedbacks = jest.fn();
    const feedbackSessionId = 101;
    const wrapper = shallow(
        <Feedbacks
            match={{params: {id: feedbackSessionId}}}
            feedbacks={feedbacks}
            getAllFeedbacks={getAllFeedbacks}
            getAllUsers={jest.fn()}
            giveFeedback={jest.fn()}
        />
    );

    expect(wrapper.state().isGivingFeedback).toBeFalsy();
  });

  it('should call get all users on component did mount', () => {
    const getAllFeedbacks = jest.fn();
    const feedbackSessionId = 101;
    const getAllUsers = jest.fn();
    const wrapper = shallow(
        <Feedbacks
            match={{params: {id: feedbackSessionId}}}
            feedbacks={feedbacks}
            getAllFeedbacks={getAllFeedbacks}
            getAllUsers={getAllUsers}
            giveFeedback={jest.fn()}
        />
    );

    expect(getAllUsers).toHaveBeenCalledWith();
  });

  it('should set give feedback pop up to true on click of button', () => {
    const getAllFeedbacks = jest.fn();
    const feedbackSessionId = 101;
    const createFeedback = jest.fn();
    const wrapper = shallow(
        <Feedbacks
            match={{params: {id: feedbackSessionId}}}
            feedbacks={feedbacks}
            getAllFeedbacks={getAllFeedbacks}
            getAllUsers={jest.fn()}
            giveFeedback={createFeedback}
        />
    );

    wrapper.find('#give-feedback-button').simulate('click');

    expect(wrapper.state().isGivingFeedback).toBeTruthy();
  });

  it('should show give feedback pop up when isGivingFeedback is true', () => {
    const getAllFeedbacks = jest.fn();
    const feedbackSessionId = 101;
    const createFeedback = jest.fn();
    const wrapper = shallow(
        <Feedbacks
            match={{params: {id: feedbackSessionId}}}
            feedbacks={feedbacks}
            getAllFeedbacks={getAllFeedbacks}
            getAllUsers={jest.fn()}
            giveFeedback={createFeedback}
        />
    );

    wrapper.find('#give-feedback-button').simulate('click');

    expect(wrapper.find(FeedbackPopUp).length).toBe(1);
  });

  it('should provide users as property to FeedbackPopUp', () => {
    const feedbackSessionId = 101;
    const users = [{id: 10}];
    const wrapper = shallow(
        <Feedbacks
            match={{params: {id: feedbackSessionId}}}
            feedbacks={[]}
            users={users}
            getAllFeedbacks={jest.fn()}
            getAllUsers={jest.fn()}
            giveFeedback={jest.fn()}
        />
    );

    wrapper.setState({isGivingFeedback: true});

    expect(wrapper.find(FeedbackPopUp).props().users).toBe(users);
  });

  it('should change pop up on call of onCancel of FeedbackPopUp', () => {
    const feedbackSessionId = 101;
    const users = [{id: 10}];
    const wrapper = shallow(
        <Feedbacks
            match={{params: {id: feedbackSessionId}}}
            feedbacks={[]}
            users={users}
            getAllFeedbacks={jest.fn()}
            getAllUsers={jest.fn()}
            giveFeedback={jest.fn()}
        />
    );

    wrapper.setState({isGivingFeedback: true});
    wrapper.find(FeedbackPopUp).props().onCancel();

    expect(wrapper.state().isGivingFeedback).toBeFalsy();
  });

  it('should call create feedback on call of onSubmit of pop up', () => {
    const feedbackSessionId = 101;
    const users = [{id: 10}];
    const giveFeedback = jest.fn();
    const wrapper = shallow(
        <Feedbacks
            match={{params: {id: feedbackSessionId}}}
            feedbacks={[]}
            users={users}
            getAllFeedbacks={jest.fn()}
            getAllUsers={jest.fn()}
            giveFeedback={giveFeedback}
        />
    );

    wrapper.setState({isGivingFeedback: true});
    const feedback = {
      description: 'description data',
      giveTo: 'user',
    };

    wrapper.find(FeedbackPopUp).props().onSubmit(feedback);
    expect(giveFeedback).toHaveBeenCalledWith(
        {...feedback, sessionId: feedbackSessionId});
  });
});
