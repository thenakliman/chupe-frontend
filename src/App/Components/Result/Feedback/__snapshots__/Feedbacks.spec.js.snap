// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="feedback-container"
>
  <div
    className="give-feedback-button-container"
  >
    <button
      className="give-feedback-button"
      id="give-feedback-button"
      onClick={[Function]}
    >
      Give Feedback
    </button>
  </div>
  <table
    id="all-feedbacks-ordered-list"
  >
    <thead>
      <tr
        className="feedbacks-table-header"
      >
        <th>
           Id
        </th>
        <th>
           description
        </th>
        <th>
           Given By
        </th>
      </tr>
    </thead>
    <tbody
      className="feedbacks-table-body"
      id="all-feedbacks-table-body-id"
    >
      <tr
        key="10101"
      >
        <td>
           
          10101
           
        </td>
        <td
          className="feedbacks-row"
        >
          retro - 1
        </td>
        <td>
           
        </td>
      </tr>
      <tr
        key="10102"
      >
        <td>
           
          10102
           
        </td>
        <td
          className="feedbacks-row"
        >
          retro - 2
        </td>
        <td>
           
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
